Tests/NIOHTTP1Tests/HTTPServerUpgradeTests.swift

```diff
@@ -379,30 +380,30 @@ private class UpgradeDelayer: TypedAndUntypedHTTPServerProtocolUpgrader {
     }
 
     public func upgrade(context: ChannelHandlerContext, upgradeRequest: HTTPRequestHead) -> EventLoopFuture<Void> {
-        self.upgradePromise = context.eventLoop.makePromise()
         upgradeRequestedPromise.succeed()
-        return self.upgradePromise!.futureResult.map { _ in }
+        return self.upgradePromise.futureResult.map { _ in }
     }
 
     public func unblockUpgrade() {
-        self.upgradePromise!.succeed(true)
+        self.upgradePromise.succeed(true)
     }
 
     func upgrade(channel: Channel, upgradeRequest: HTTPRequestHead) -> EventLoopFuture<Bool> {
-        self.upgradePromise = channel.eventLoop.makePromise()
         self.upgradeRequestedPromise.succeed()
-        return self.upgradePromise!.futureResult
+        return self.upgradePromise.futureResult
     }
 }
 
-private class UpgradeResponseDelayer: HTTPServerProtocolUpgrader {
+private final class UpgradeResponseDelayer: HTTPServerProtocolUpgrader, Sendable {
     let supportedProtocol: String
     let requiredUpgradeHeaders: [String] = []
 
-    private var context: ChannelHandlerContext?
-    private let buildUpgradeResponseHandler: () -> EventLoopFuture<Void>
+    private let buildUpgradeResponseHandler: @Sendable () -> EventLoopFuture<Void>
 
-    public init(forProtocol `protocol`: String, buildUpgradeResponseHandler: @escaping () -> EventLoopFuture<Void>) {
+    public init(
```

Comment:
Unrelated to your change but shouldn't this be `private` or `internal`? The class is `private`.